# Introduction

This is my first dive into a solo analysis using linear regression and US census data. My hope is to answer the following questions:
1. Can I predict crimes of rape, arson, vehicle theft, larceny, burglary, robbery, and aggravated assault using demographic data?
2. Can I determine which demographic fields are the most useful for such predictions? 
3. Can I apply the models to predict crime rates for some of the cities I'm considering moving to?

# Imports and Datasets

import the required modules and read in the datasets.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
import seaborn as sns
%matplotlib inline

#DataSets for analysis
#source: https://hub.arcgis.com/datasets/6afda9afb9f94e27adf39d36e119eb29_2
demo_df= pd.read_csv('./Census_demographics_by_County.csv')
#source: https://hub.arcgis.com/datasets/utah::us-states
demo_state_df = pd.read_csv("./US_States.csv")
#source: https://www.kaggle.com/mikejohnsonjr/united-states-crime-rates-by-county
crime_df = pd.read_csv('./crime_data_w_population_and_crime_rate.csv')

# Reviewing The Datasets

### Crime_df

The _crime_df_ dataset contains the crimes reported by counties in 2017. We are interested in predicting the crimes:
1. Rape (RAPE)
2. Arson (ARSON)
3. Motor Vehicle Theft (MVTHEFT)
4. Larceny (LARCENY)
5. Burglary (BURGLRY)
6. Robbery (Robbery)
7. Aggravated Assult (AGASSLT)


In [2]:
crime_df.head(2)

county_name  crime_rate_per_100000  index  EDITION  PART  IDNO  \
0     St. Louis city, MO            1791.995377      1        1     4  1612   
1  Crittenden County, AR            1754.914968      2        1     4   130   

   CPOPARST  CPOPCRIM  AG_ARRST  AG_OFF  ...  RAPE  ROBBERY  AGASSLT  BURGLRY  \
0    318667    318667        15      15  ...   200     1778     3609     4995   
1     50717     50717         4       4  ...    38      165      662     1482   

   LARCENY  MVTHEFT  ARSON  population  FIPS_ST  FIPS_CTY  
0    13791     3543    464      318416       29       510  
1     1753      189     28       49746        5        35  

[2 rows x 24 columns]

### Demo_df

The demo_df dataset contains US Census data that will be used as additional features to try to predict crime with.

In [3]:
demo_df.head(2)

OBJECTID  STATE  STCN        CNTYNAME    AREALAND  AREAWATER  TOTALPOP  \
0         1      1  1001  Autauga County  1539584444   25773561     55136   
1         2      1  1005  Barbour County  2291820953   50864677     27119   

     POP_DEN  NUM_MINORITY  PCT_MINORITY  ...  EMP_STAT_NOT_IN_LF  \
0  92.753326         13103     23.764872  ...               15561   
1  30.647195         14614     53.888418  ...               11086   

   PCT_EMP_STAT_NOT_IN_LF  Shape_Length  Shape_Area  FBPOP  EUROPE   ASIA  \
0               36.603782      2.052620    0.150260  772.0   141.0  150.0   
1               50.736842      2.566611    0.223256  769.0    14.0  169.0   

   AFRICA  OCEANIA  AMERICAS  
0     0.0    119.0     362.0  
1     0.0      0.0     586.0  

[2 rows x 266 columns]

### Demo_state_df

The demo_state_df dataset is used to join the demo_df and the crime_df. While, both demo_df and crime_df have the county and the state, the demo_df uses a numeric designation for the state while crime_df uses the state abbreviation. 

In [4]:
demo_state_df.head(2)

OBJECTID_1  OBJECTID STATE_NAME  STATE_FIPS STATE_ABBR  SHAPE_Length  \
0           1       NaN    Alabama           1         AL     20.554796   
1           2       NaN     Alaska           2         AK    318.661159   

   SHAPE_Area  
0   12.881818  
1  167.784052

# Data Wrangling

Start by adding a state abbreviation field to the demo_df dataset to join with crime_df later.

In [5]:
#Use demo_state_df to recover the states in demo_df

#Match on the state numeric code
for name in range(len(demo_df['STATE'])):
    for stateCode in range(len(demo_state_df['STATE_FIPS'])):
        if(demo_df['STATE'][name] == demo_state_df['STATE_FIPS'][stateCode]):
            demo_df['STATE'][name] = demo_state_df['STATE_ABBR'][stateCode]

<ipython-input-5-1dfa1662d4ef>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo_df['STATE'][name] = demo_state_df['STATE_ABBR'][stateCode]
/home/ada/pythonJupyter/jupyterenvironment/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [6]:
#Split string on comma and turn into a list. Needed for matching later.
for name in range(len(crime_df['county_name'])):
    crime_df['county_name'][name] = crime_df['county_name'][name].split(",")
    crime_df['county_name'][name][1] = crime_df['county_name'][name][1].strip()

<ipython-input-6-3bc2aae1c169>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crime_df['county_name'][name] = crime_df['county_name'][name].split(",")


In [7]:
# Add columns for matching
crime_df.insert(0,"JoinColumn", [np.nan] * crime_df.shape[0], True )
demo_df.insert(0,"JoinColumn", [np.nan] * demo_df.shape[0], True )

In [8]:
# Match based on the county name and the state
numEqual = 0
for x in range(len(crime_df['county_name'])):
    for y in range(len(demo_df['CNTYNAME'])):
        if crime_df['county_name'][x][0] == demo_df['CNTYNAME'][y]:
            if(crime_df['county_name'][x][1] == demo_df['STATE'][y]):
                numEqual += 1
                crime_df['JoinColumn'][x] = numEqual
                demo_df['JoinColumn'][y] = numEqual

<ipython-input-8-bb558d746821>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crime_df['JoinColumn'][x] = numEqual
<ipython-input-8-bb558d746821>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo_df['JoinColumn'][y] = numEqual


### Wrangling Results

The prints below show that we have 3135 matches and had to drop a little under 100 counties from the demo_df dataset. 

In [9]:
#Show the number of matches                
print(numEqual)
print(crime_df.shape)
print(demo_df.shape)
#Drop Columns that did not have a match
crime_df_dropped = crime_df.dropna(subset=['JoinColumn'], how = 'any')
demo_df_dropped = demo_df.dropna(subset=['JoinColumn'], how = 'any')
#Show the number of matches in each data frame
#Note that they are the same size, there were not any duplicate matches
print(crime_df_dropped.shape)
print(demo_df_dropped.shape)

3135
(3136, 25)
(3220, 267)
(3135, 25)
(3135, 267)


# Aggregating Data and Imputation

Next, join the crime_df and demo_df datasets based on the matching column.

Impute values for the cities:
1. Columbia, Howard County
2. Odenton, Anne Arundel County
3. Glen Burnie, Anne Arundel County
4. Hanover, Anne Arundel County

Then drop the identification based fields.

### On Imputation Methodology

I could not find a dataset for the cities that consisted of the same fields found in the demographic demo_df dataset. So I extract the rows for the counties the cites are located in then reduce the numeric fields by the difference in population. For example, if Howard County had 1000 Asian females and Columbia had 1/4 of the population of Howard County, then the value of the field for Asian females would be 250.


It basically assumes that the counties have a uniform demographic distribution. While that is is not a good assumption, the result is for personal estimation only, so I accept the inaccuracies associated with the assumption :)

In [10]:

#Columns that are not to be used in the modeling and will be dropped.
dropList = ['JoinColumn','county_name', 'EDITION', 'PART', 'IDNO', 'INDEX', 'FIPS_ST', 'FIPS_CTY','OBJECTID', 'STATE', 'STCN', 'CNTYNAME']

#Join on the matches
joinedDF = crime_df_dropped.merge(demo_df_dropped, on='JoinColumn')

#Subset by county based on potential living city locations
columbia = joinedDF[joinedDF['STATE'] =='MD'].copy()
columbia = columbia[columbia['CNTYNAME'] == "Howard County"]

odenton = joinedDF[joinedDF['STATE'] =='MD'].copy()
odenton = odenton[odenton['CNTYNAME'] == "Anne Arundel County"]

glen_burnie = joinedDF[joinedDF['STATE'] =='MD'].copy()
glen_burnie = glen_burnie[glen_burnie['CNTYNAME'] == "Anne Arundel County"]

hanover = joinedDF[joinedDF['STATE'] =='MD'].copy()
hanover = hanover[hanover['CNTYNAME'] == "Anne Arundel County"]

#Drop unneccessary columns
columbia = columbia.drop(dropList, axis =1)
odenton = odenton.drop(dropList, axis =1)
glen_burnie = glen_burnie.drop(dropList, axis =1)
hanover = hanover.drop(dropList, axis =1)
joinedDF = joinedDF.drop(dropList, axis =1)

# Population in 2019
# Used for imputation
howardCountyPop = 325690
columbiaPop = 103991
anneArundelPop = 579234
odentonPop = 41846
glen_burniePop = 69872
hanoverPop = 12952

#Impute the numeric fields based on the population proportion
def imputeSubset(df, portion):
    for column in df.columns:
        if np.issubdtype(df.dtypes[column], np.number):
                df[column] = df[column] * portion

                
#Impute values for the cities from their county
imputeSubset(columbia, columbiaPop / howardCountyPop)
imputeSubset(odenton, odentonPop / anneArundelPop)
imputeSubset(glen_burnie, glen_burniePop / anneArundelPop)
imputeSubset(hanover, hanoverPop / anneArundelPop)

# Modeling

The function defined below is used to create models of:

1. Rape (RAPE)
2. Arson (ARSON)
3. Motor Vehicle Theft (MVTHEFT)
4. Larceny (LARCENY)
5. Burglary (BURGLRY)
6. Robbery (Robbery)
7. Aggravated Assult (AGASSLT)

for the cities: 

1. Columbia, Howard County
2. Odenton, Anne Arundel County
3. Glen Burnie, Anne Arundel County
4. Hanover, Anne Arundel County

In [11]:
def model(parameter):
    X = joinedDF.drop([parameter], axis = 1)
    y = joinedDF[parameter]
    #Split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .30, random_state=101011)
    # train
    lm_model = LinearRegression(normalize=True) # Instantiate
    lm_model.fit(X_train, y_train) #Fit
    # predict
    y_test_preds = lm_model.predict(X_test)
    y_train_preds = lm_model.predict(X_train)
    #score
    test_score = r2_score(y_test, y_test_preds)
    train_score = r2_score(y_train, y_train_preds)
    print("The rsquared on the training data was {}.  The rsquared on the test data was {}.".format(train_score, test_score))
    
    #For review of the model's performance
    coefs_df = pd.DataFrame()
    coefs_df['est_int'] = X_train.columns
    coefs_df['coefs'] = lm_model.coef_
    coefs_df['abs_coefs'] = np.abs(lm_model.coef_)
    coefs_df = coefs_df.sort_values('abs_coefs', ascending=False)
    compareDF = pd.DataFrame({'Actual': y_test, 'Predicted': y_test_preds})
    
    #Subset the city data based on the passed parameter
    columbiaDropped = columbia.drop([parameter], axis = 1)
    odentonDropped =odenton.drop([parameter], axis = 1)
    glen_burnieDropped = glen_burnie.drop([parameter], axis = 1)
    hanoverDropped = hanover.drop([parameter], axis = 1)
    
    #Attempt to predict the parameter using the model
    columbiaVal = lm_model.predict(columbiaDropped)[0] 
    odentonVal = lm_model.predict(odentonDropped)[0] 
    glen_burnieVal = lm_model.predict(glen_burnieDropped)[0] 
    hanoverVal = lm_model.predict(hanoverDropped)[0] 
    
    return [[columbiaVal, odentonVal,glen_burnieVal,hanoverVal], coefs_df, compareDF] 

# Results

The following cells call the model function on the parameter of interest then print the rsquared score, the most impactful field, and a comparison of the predicted and actual values

In [12]:
results = pd.DataFrame({'City' : ["Columbia", "Odenton", "Glen Burnie", "Hanover"]})
results.set_index('City')
print(results)

          City
0     Columbia
1      Odenton
2  Glen Burnie
3      Hanover


In [13]:
result = model("RAPE")
results.insert(1, "RAPE",result[0])
result[2]

The rsquared on the training data was 0.9737776942603972.  The rsquared on the test data was 0.7162012091176642.


Actual   Predicted
2524       0   -0.218750
3102       0  -10.609375
82        31   26.984375
1573      94  109.734375
2632       1   -0.343750
...      ...         ...
2501       7    1.562500
966        8    7.906250
439      289  318.546875
2943       0   -6.781250
2835       2    6.593750

[941 rows x 2 columns]

In [14]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  1.251427e+12  1.251427e+12
259      PCT_EMP_STAT_IN_LF  1.251427e+12  1.251427e+12
171           PCT_HLI_IE_LI  1.507803e+11  1.507803e+11
173          PCT_HLI_IE_NLI  9.297506e+10  9.297506e+10
59             PCT_AGE_GT17 -8.781943e+10  8.781943e+10
53             PCT_AGE_LT18 -8.781943e+10  8.781943e+10
169              PCT_HLI_IE -8.753956e+10  8.753956e+10
181           PCT_HLI_OTHER  8.232970e+10  8.232970e+10
185       PCT_HLI_OTHER_NLI -7.689419e+10  7.689419e+10
148           PCT_LAN_ENG_W -6.921693e+10  6.921693e+10
150          PCT_LAN_ENG_NW -5.860229e+10  5.860229e+10
152          PCT_LAN_ENG_NA -5.860229e+10  5.860229e+10
159             PCT_LINGISO -5.780523e+10  5.780523e+10
175             PCT_HLI_API  5.022252e+10  5.022252e+10
179         PCT_HLI_API_NLI -4.478702e+10  4.478702e+10
156            PCT_ENG_LTVW  4.334871e+10  4.334871e+10
144         PCT_LAN_NON_ENG  3.643051e+10  3.643051e+10
165      PCT_HLI_SPANISH_LI  3.236567e+10  3.236567e+10
163         PCT_HLI_SPANISH  3.087507e+10  3.087507e+10
21             PCT_MINORITY -2.851222e+10  2.851222e+10

In [15]:
result = model("ARSON")
results.insert(1, "ARSON",result[0])
result[2]
#Predictions for ARSON are very inaccuarate

The rsquared on the training data was 0.9539673580763047.  The rsquared on the test data was 0.3397538395428641.


Actual  Predicted
2524       0   -2.40625
3102       0   -3.37500
82        13   11.84375
1573      39   94.84375
2632       1   -1.50000
...      ...        ...
2501       3    0.78125
966        1    8.06250
439      169  150.37500
2943       0   -4.56250
2835       5    5.87500

[941 rows x 2 columns]

In [16]:
result[1].head(20)

est_int         coefs     abs_coefs
259      PCT_EMP_STAT_IN_LF -1.740746e+12  1.740746e+12
269  PCT_EMP_STAT_NOT_IN_LF -1.740746e+12  1.740746e+12
179         PCT_HLI_API_NLI  2.412376e+11  2.412376e+11
175             PCT_HLI_API -2.234421e+11  2.234421e+11
171           PCT_HLI_IE_LI -1.938352e+11  1.938352e+11
183        PCT_HLI_OTHER_LI -1.578975e+11  1.578975e+11
177          PCT_HLI_API_LI  1.282647e+11  1.282647e+11
159             PCT_LINGISO  1.129729e+11  1.129729e+11
169              PCT_HLI_IE  9.865786e+10  9.865786e+10
173          PCT_HLI_IE_NLI -8.086235e+10  8.086235e+10
165      PCT_HLI_SPANISH_LI -7.285435e+10  7.285435e+10
148           PCT_LAN_ENG_W  7.136630e+10  7.136630e+10
181           PCT_HLI_OTHER  6.272007e+10  6.272007e+10
150          PCT_LAN_ENG_NW  5.691490e+10  5.691490e+10
152          PCT_LAN_ENG_NA  5.691490e+10  5.691490e+10
185       PCT_HLI_OTHER_NLI -4.492456e+10  4.492456e+10
156            PCT_ENG_LTVW -4.013858e+10  4.013858e+10
167     PCT_HLI_SPANISH_NLI  4.011854e+10  4.011854e+10
144         PCT_LAN_NON_ENG -3.311872e+10  3.311872e+10
146          PCT_LAN_ENG_VW  3.122772e+10  3.122772e+10

In [17]:
result = model("MVTHEFT")
results.insert(1, "MVTHEFT",result[0])
result[2]

The rsquared on the training data was 0.9999999521064714.  The rsquared on the test data was 0.9999998983422953.


Actual    Predicted
2524       4     4.035400
3102       0     0.294189
82       119   118.950195
1573     464   464.180664
2632       0    -0.067383
...      ...          ...
2501      14    14.032715
966       34    33.833008
439     3967  3967.805664
2943       1     0.883057
2835      31    30.875000

[941 rows x 2 columns]

In [18]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  9.995406e+09  9.995406e+09
259      PCT_EMP_STAT_IN_LF  9.995406e+09  9.995406e+09
177          PCT_HLI_API_LI -3.694731e+09  3.694731e+09
179         PCT_HLI_API_NLI -3.188915e+09  3.188915e+09
175             PCT_HLI_API  3.086010e+09  3.086010e+09
59             PCT_AGE_GT17  2.967552e+09  2.967552e+09
53             PCT_AGE_LT18  2.967552e+09  2.967552e+09
183        PCT_HLI_OTHER_LI -1.833522e+09  1.833522e+09
169              PCT_HLI_IE -1.369639e+09  1.369639e+09
152          PCT_LAN_ENG_NA  1.360300e+09  1.360300e+09
150          PCT_LAN_ENG_NW  1.360300e+09  1.360300e+09
185       PCT_HLI_OTHER_NLI -1.327706e+09  1.327706e+09
173          PCT_HLI_IE_NLI  1.266735e+09  1.266735e+09
181           PCT_HLI_OTHER  1.224802e+09  1.224802e+09
63              PCT_FEMALES  1.055704e+09  1.055704e+09
61                PCT_MALES  1.055704e+09  1.055704e+09
154         PCT_LAN_ENG_LTW -1.034292e+09  1.034292e+09
171           PCT_HLI_IE_LI  7.609186e+08  7.609186e+08
146          PCT_LAN_ENG_VW  7.530075e+08  7.530075e+08
261  PCT_EMP_STAT_IN_CIV_LF  6.418708e+08  6.418708e+08

In [19]:
result = model("LARCENY")
results.insert(1, "LARCENY",result[0])
result[2]

The rsquared on the training data was 0.9999999979573293.  The rsquared on the test data was 0.999999997083696.


Actual     Predicted
2524      21     21.036621
3102       2      2.278809
82      1600   1599.951172
1573    7010   7010.199463
2632      42     41.932373
...      ...           ...
2501     169    169.031494
966      610    609.831055
439    20856  20856.776123
2943      17     16.878906
2835     804    803.872559

[941 rows x 2 columns]

In [20]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  1.021766e+10  1.021766e+10
259      PCT_EMP_STAT_IN_LF  1.021766e+10  1.021766e+10
177          PCT_HLI_API_LI -3.750320e+09  3.750320e+09
179         PCT_HLI_API_NLI -3.239929e+09  3.239929e+09
175             PCT_HLI_API  3.136848e+09  3.136848e+09
59             PCT_AGE_GT17  2.973672e+09  2.973672e+09
53             PCT_AGE_LT18  2.973672e+09  2.973672e+09
183        PCT_HLI_OTHER_LI -1.840216e+09  1.840216e+09
169              PCT_HLI_IE -1.367656e+09  1.367656e+09
152          PCT_LAN_ENG_NA  1.353940e+09  1.353940e+09
150          PCT_LAN_ENG_NW  1.353940e+09  1.353940e+09
185       PCT_HLI_OTHER_NLI -1.329825e+09  1.329825e+09
173          PCT_HLI_IE_NLI  1.264575e+09  1.264575e+09
181           PCT_HLI_OTHER  1.226745e+09  1.226745e+09
154         PCT_LAN_ENG_LTW -1.034873e+09  1.034873e+09
171           PCT_HLI_IE_LI  7.541841e+08  7.541841e+08
146          PCT_LAN_ENG_VW  7.513078e+08  7.513078e+08
261  PCT_EMP_STAT_IN_CIV_LF  6.414967e+08  6.414967e+08
267      PCT_EMP_STAT_IN_AF  6.414967e+08  6.414967e+08
165      PCT_HLI_SPANISH_LI -5.669825e+08  5.669825e+08

In [21]:
result = model("BURGLRY")
results.insert(1, "BURGLRY",result[0])
result[2]

The rsquared on the training data was 0.9999999816071707.  The rsquared on the test data was 0.9999999793052033.


Actual    Predicted
2524       4     4.033447
3102       0     0.291016
82       734   733.950928
1573    2155  2155.163330
2632      13    12.932373
...      ...          ...
2501     100   100.031738
966      235   234.832275
439     9180  9180.809326
2943       7     6.881104
2835     265   264.871094

[941 rows x 2 columns]

In [22]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  1.030303e+10  1.030303e+10
259      PCT_EMP_STAT_IN_LF  1.030303e+10  1.030303e+10
177          PCT_HLI_API_LI -3.760505e+09  3.760505e+09
179         PCT_HLI_API_NLI -3.250798e+09  3.250798e+09
175             PCT_HLI_API  3.147183e+09  3.147183e+09
59             PCT_AGE_GT17  2.969849e+09  2.969849e+09
53             PCT_AGE_LT18  2.969849e+09  2.969849e+09
61                PCT_MALES -1.885949e+09  1.885949e+09
63              PCT_FEMALES -1.885949e+09  1.885949e+09
183        PCT_HLI_OTHER_LI -1.833829e+09  1.833829e+09
169              PCT_HLI_IE -1.360925e+09  1.360925e+09
152          PCT_LAN_ENG_NA  1.352245e+09  1.352245e+09
150          PCT_LAN_ENG_NW  1.352245e+09  1.352245e+09
185       PCT_HLI_OTHER_NLI -1.324122e+09  1.324122e+09
173          PCT_HLI_IE_NLI  1.257310e+09  1.257310e+09
181           PCT_HLI_OTHER  1.220508e+09  1.220508e+09
154         PCT_LAN_ENG_LTW -1.033407e+09  1.033407e+09
146          PCT_LAN_ENG_VW  7.508722e+08  7.508722e+08
171           PCT_HLI_IE_LI  7.476034e+08  7.476034e+08
261  PCT_EMP_STAT_IN_CIV_LF  6.400322e+08  6.400322e+08

In [23]:
result = model("AGASSLT")
results.insert(1, "AGASSLT",result[0])
result[2]

The rsquared on the training data was 0.9858276515273506.  The rsquared on the test data was 0.7630907938511257.


Actual  Predicted
2524       2    -74.875
3102       0    128.625
82       394    412.875
1573     895   1062.500
2632       8     11.500
...      ...        ...
2501       7      9.250
966       52     38.500
439     2274    966.375
2943       1     51.375
2835      36     11.750

[941 rows x 2 columns]

In [24]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF -6.239436e+12  6.239436e+12
259      PCT_EMP_STAT_IN_LF -6.239436e+12  6.239436e+12
59             PCT_AGE_GT17 -1.605918e+12  1.605918e+12
53             PCT_AGE_LT18 -1.605918e+12  1.605918e+12
183        PCT_HLI_OTHER_LI  1.012390e+12  1.012390e+12
63              PCT_FEMALES -9.243892e+11  9.243892e+11
61                PCT_MALES -9.243892e+11  9.243892e+11
185       PCT_HLI_OTHER_NLI  7.759634e+11  7.759634e+11
181           PCT_HLI_OTHER -6.766975e+11  6.766975e+11
169              PCT_HLI_IE  5.823059e+11  5.823059e+11
173          PCT_HLI_IE_NLI -4.830400e+11  4.830400e+11
177          PCT_HLI_API_LI  3.845554e+11  3.845554e+11
267      PCT_EMP_STAT_IN_AF  3.350314e+11  3.350314e+11
261  PCT_EMP_STAT_IN_CIV_LF  3.350314e+11  3.350314e+11
154         PCT_LAN_ENG_LTW -3.280026e+11  3.280026e+11
152          PCT_LAN_ENG_NA  3.274594e+11  3.274594e+11
150          PCT_LAN_ENG_NW  3.274594e+11  3.274594e+11
163         PCT_HLI_SPANISH  3.242760e+11  3.242760e+11
171           PCT_HLI_IE_LI -2.466134e+11  2.466134e+11
159             PCT_LINGISO -2.364266e+11  2.364266e+11

In [25]:
result = model("ROBBERY")
results.insert(1, "ROBBERY",result[0])
result[2]

The rsquared on the training data was 0.9969916270011239.  The rsquared on the test data was 0.823755330025302.


Actual    Predicted
2524       1   -12.359375
3102       0    51.523438
82        49    72.218750
1573     280   719.968750
2632       1    -7.226562
...      ...          ...
2501       1    14.585938
966       14    21.843750
439      910  1158.609375
2943       1    19.203125
2835       5    -5.460938

[941 rows x 2 columns]

In [26]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF -1.141678e+12  1.141678e+12
259      PCT_EMP_STAT_IN_LF -1.141678e+12  1.141678e+12
177          PCT_HLI_API_LI -4.676747e+11  4.676747e+11
171           PCT_HLI_IE_LI -3.980697e+11  3.980697e+11
183        PCT_HLI_OTHER_LI -3.858850e+11  3.858850e+11
61                PCT_MALES  3.755275e+11  3.755275e+11
63              PCT_FEMALES  3.755275e+11  3.755275e+11
165      PCT_HLI_SPANISH_LI -3.138486e+11  3.138486e+11
159             PCT_LINGISO  2.749814e+11  2.749814e+11
154         PCT_LAN_ENG_LTW  2.439053e+11  2.439053e+11
179         PCT_HLI_API_NLI -1.926933e+11  1.926933e+11
175             PCT_HLI_API  1.757614e+11  1.757614e+11
148           PCT_LAN_ENG_W  1.368417e+11  1.368417e+11
156            PCT_ENG_LTVW -1.330406e+11  1.330406e+11
53             PCT_AGE_LT18  1.274358e+11  1.274358e+11
59             PCT_AGE_GT17  1.274358e+11  1.274358e+11
173          PCT_HLI_IE_NLI -1.230883e+11  1.230883e+11
185       PCT_HLI_OTHER_NLI -1.109036e+11  1.109036e+11
152          PCT_LAN_ENG_NA -1.070636e+11  1.070636e+11
150          PCT_LAN_ENG_NW -1.070636e+11  1.070636e+11

In [27]:
results.sum(axis = 1)

0    6.612362e+14
1    9.012209e+14
2    8.542202e+14
3    9.496773e+14
dtype: float64

# Conclusion

### Model Evaluation
The results for Arson has the lowest R squared score of 0.3397538395428641, which is too low to comfortably use conclusively. Additionally, Burglary, Motor Vehicle Theft, and Larceny had very high scores. They are so high that I'm concerned I overfitted.

The remaining scores were:
1. Rape: 0.7162012091176642
2. Aggravated Assult: 0.7630907938511257
3. Robbery: 0.823755330025302

So I will use the rape, aggravated assult, and robbery fields to help me decide where I should live.

### Influential Demographic Fields 

The following fields seemed to have the largest influence on the model:
1. PCT_EMP_STAT_NOT_IN_LF (those not employed in the labor force?)
2. PCT_EMP_STAT_IN_LF	(those who are employed in the labor force?)
3. PCT_HLI_API_LI (Asian pacific Islanders who are linguistically isolated?)
4. PCT_HLI_IE_LI (no clue)
5. PCT_AGE_LT18 (Less than 18 years old)
6. PCT_AGE_GT17 (Greater than 17 years old?)

I stuggled to find documentation on these fields and the phrases in parentheses are the best descriptions I could come up with. 

__NOTE:__ _These fields are used for prediction and do not neccessarily mean causation_. I am not saying that non-english speaking people are the cause of crime in this county

In [28]:
violent = results.drop(['BURGLRY', 'LARCENY', 'MVTHEFT', 'ARSON'], axis = 1)
#Sum for 'RAPE', 'AGASSLT', 'ROBBERY'
violent.sum(axis = 1)

0    5.443797e+14
1    7.419532e+14
2    7.032587e+14
3    7.818462e+14
dtype: float64

In [29]:
prop = results.drop(['RAPE', 'AGASSLT', 'ROBBERY', 'ARSON'], axis = 1)
#Sum for 'BURGLRY', 'LARCENY', 'MVTHEFT'
prop.sum(axis = 1)

0   -2.883787e+12
1   -3.930409e+12
2   -3.725429e+12
3   -4.141737e+12
dtype: float64

In [30]:
all = results.drop(['ARSON'], axis = 1)
all.sum(axis = 1)

0    5.414959e+14
1    7.380228e+14
2    6.995332e+14
3    7.777044e+14
dtype: float64

For the three cities in Anne Arundel County, it seems that the smaller the city, the higher the crime?

This is not mirrored by online sources and is probably indicative that my method of imputation is a lot worse than I thought.